File: Microservices/IoT.House.Automation.Microservices.Arduino/IoT.House.Automation.Microservices.Arduino.Infra/RabbitMQ/EventBus/RabbitMQEventBus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using IoT.House.Automation.Microservices.Arduino.Application.Interfaces;
using IoT.House.Automation.Microservices.Arduino.Application.MessageBroker.Events;
using IoT.House.Automation.Microservices.Arduino.Application.Util;
using IoT.House.Automation.Microservices.Arduino.Infra.RabbitMQ.Connection;
using IoT.House.Automation.Microservices.Arduino.Infra.RabbitMQ.Subscription;
using Newtonsoft.Json;
using Polly;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace IoT.House.Automation.Microservices.Arduino.Infra.RabbitMQ.EventBus
{
    public class RabbitMQEventBus : IEventBus, IDisposable
    {
        private readonly IServiceProvider _provider;
        private readonly PersisterConnection _connection;
        private readonly SubscriptionManager _subscriptionManager;
        private readonly IModel _consumerChannel;

        public RabbitMQEventBus(IServiceProvider provider, PersisterConnection connection)
        {
            IfHasNullThrowException(provider, connection);

            _provider = provider;
            _connection = connection;

            _subscriptionManager = new SubscriptionManager();
            _subscriptionManager.OnEventRemoved += OnSubscriptionManagerEventRemoved;
            _subscriptionManager.OnEventAdded += OnSubscriptionManagerEventAdded;

            _consumerChannel = CreateConsumerChannel();

        }

        private void OnSubscriptionManagerEventAdded(object _, string eventName)
        {
            using (var channel = _connection.CreateModel())
            {
                channel.QueueDeclare(eventName, true, false, false);

                channel.QueueBind(
                    queue: eventName,
                    exchange: _connection.Config.Exchange,
                    routingKey: eventName
                );

                CreateConsumer(_consumerChannel, eventName);
            }
        }

        private void OnSubscriptionManagerEventRemoved(object _, string eventName)
        {
            using (var channel = _connection.CreateModel())
            {
                channel.QueueUnbind(
                    queue: eventName,
                    exchange: _connection.Config.Exchange,
                    routingKey: eventName
                );

                _consumerChannel.BasicCancel(eventName);

                if (!_subscriptionManager.IsEmpty) return;
                _consumerChannel.Close();
            }
        }

        public void Publish(Event @event)
        {
            var policy = Policy.Handle<Exception>()
                .WaitAndRetry(5, retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)), (ex, time) =>
                {
                });

            using (var channel = _connection.CreateModel())
            {
                var eventName = @event.GetType()
                    .Name;

                channel.ExchangeDeclare(exchange: _connection.Config.Exchange, type: _connection.Config.ExchangeType);

                var settings = NewtonsoftJsonUtil.UseCustomNewtonsoftSettings();

                var message = JsonConvert.SerializeObject(@event, settings);
                var body = Encoding.UTF8.GetBytes(message);

                policy.Execute(() =>
                {
                    channel.BasicPublish(exchange: _connection.Config.Exchange,
                        routingKey: eventName,
                        basicProperties: null,
                        body: body);
                });
            }
        }

        public void Subscribe<TEvent, TEventHandler>() where TEvent : Event where TEventHandler : IEventHandler<TEvent>
        {
            _subscriptionManager.AddSubscription<TEvent, TEventHandler>();
        }

        public void Unsubscribe<TEvent, TEventHandler>() where TEvent : Event where TEventHandler : IEventHandler<TEvent>
        {
            _subscriptionManager.RemoveSubscription<TEvent, TEventHandler>();
        }

        private IModel CreateConsumerChannel()
        {

            var channel = _connection.CreateModel();
            channel.ExchangeDeclare(exchange: _connection.Config.Exchange, type: _connection.Config.ExchangeType);
            return channel;
        }

        private void CreateConsumer(IModel channel, string queueName)
        {
            var consumer = new EventingBasicConsumer(channel);

            consumer.Received += async (model, ea) =>
            {
                var eventName = ea.RoutingKey;
                var message = Encoding.UTF8.GetString(ea.Body.ToArray());

                await HandleEvent(eventName, message, ea.DeliveryTag, channel);
            };

            channel.BasicConsume(queue: queueName,
                autoAck: false,
                consumer: consumer,
                consumerTag: queueName);

            channel.CallbackException += (sender, ea) =>
            {
                _consumerChannel.Dispose();
            };
        }

        private async Task HandleEvent(string eventName, string message, ulong deliveryTag, IModel channel)
        {
            if (!_subscriptionManager.HasSubscriptionsForEvent(eventName)) return;

            var subscriptions = _subscriptionManager.GetHandlersForEvent(eventName);

            foreach (var subscription in subscriptions)
            {
                var result= subscription.Handle(message, _provider);

                if (result.IsFaulted || result.IsCanceled)
                {
                    channel.BasicNack(deliveryTag, false, true);
                }
                else
                {
                    channel.BasicAck(deliveryTag, false);
                }
            }

        }

        private void IfHasNullThrowException(params object[] args)
        {
            args.Any(o =>
            {
                if (o == default) throw new ArgumentNullException(nameof(o));
                return false;
            });
        }

        public void Dispose()
        {
            _consumerChannel?.Dispose();
        }

    }
}

File: Microservices/IoT.House.Automation.Microservices.Mobile/infrastructure/src/main/java/iot/house/automation/microservices/mobile/infra/rabbitMQ/eventbus/RabbitMQEventBus.java
package iot.house.automation.microservices.mobile.infra.rabbitMQ.eventbus;

import com.google.gson.Gson;
import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.DefaultConsumer;
import com.rabbitmq.client.Envelope;
import iot.house.automation.microservices.mobile.application.interfaces.EventBus;
import iot.house.automation.microservices.mobile.application.interfaces.EventHandler;
import iot.house.automation.microservices.mobile.application.messaging.events.BaseEvent;
import iot.house.automation.microservices.mobile.infra.rabbitMQ.connection.PersistedConnection;
import iot.house.automation.microservices.mobile.infra.rabbitMQ.subscription.SubscriptionManager;
import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.net.PortUnreachableException;
import java.net.SocketException;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.concurrent.TimeoutException;


@Component
public class RabbitMQEventBus implements EventBus {
    private final ApplicationContext context;
    private final PersistedConnection connection;
    private final SubscriptionManager subscriptionManager;
    private final Channel consumerChannel;
    private final Gson gson;

    @Autowired
    public RabbitMQEventBus(ApplicationContext context, PersistedConnection connection) {
        this.context = context;
        this.connection = connection;
        this.gson = new Gson();
        this.subscriptionManager = new SubscriptionManager();

        subscriptionManager.setOnEventAdded(s -> onSubscriptionManagerEventAdded(s));
        subscriptionManager.setOnEventRemoved(s -> onSubscriptionManagerEventRemoved(s));

        this.consumerChannel = createConsumerChannel();
    }

    @Override
    public void publish(BaseEvent event) {
        var policy = new RetryPolicy<>()
                .handle(Exception.class)
                .withMaxRetries(5)
                .withDelay(Duration.ofMinutes(1));

        try(var channel = connection.createChannel()) {

            var eventName = event.getType();
            var message = gson.toJson(event);
            var body = message.getBytes();

            Failsafe.with(policy).run(() -> {
                channel.basicPublish(connection.getConfig().getPublishExchange(), eventName, null, body);
            });


        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public <TEvent extends BaseEvent, TEventHandler extends EventHandler> void subscribe(Class<TEvent> eventType, Class<TEventHandler> handlerType) {
        subscriptionManager.addSubscription(eventType, handlerType);
    }

    @Override
    public <TEvent extends BaseEvent, TEventHandler extends EventHandler> void unsubscribe(Class<TEvent> eventType, Class<TEventHandler> handlerType) {
        subscriptionManager.removeSubscription(eventType, handlerType);
    }

    @SneakyThrows
    private Channel createConsumerChannel() {
        var channel = connection.createChannel();
        channel.exchangeDeclare(connection.getConfig().getPublishExchange(), connection.getConfig().getExchangeType(), true);
        return channel;
    }

    private void onSubscriptionManagerEventAdded(String eventName)
    {
        try(var channel = connection.createChannel()) {

            channel.queueDeclare(eventName, true, false, false, null);
            channel.queueBind(eventName, connection.getConfig().getListenerExchange(), eventName);
            createConsumer(consumerChannel, eventName);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void onSubscriptionManagerEventRemoved(String eventName)
    {
        try(var channel = connection.createChannel()) {

            channel.queueUnbind(eventName, connection.getConfig().getListenerExchange(), eventName);
            consumerChannel.basicCancel(eventName);

            if(!subscriptionManager.isEmpty()) return;

            consumerChannel.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SneakyThrows
    private void createConsumer(Channel channel, String eventName) {
        boolean autoAck = false;
        channel.basicConsume(eventName, false, eventName,
                new DefaultConsumer(channel) {
                    @Override
                    public void handleDelivery(String consumerTag,
                                               Envelope envelope,
                                               AMQP.BasicProperties properties,
                                               byte[] body)
                            throws IOException
                    {
                        var routingKey = envelope.getRoutingKey();
                        var deliveryTag = envelope.getDeliveryTag();
                        var message = new String(body, StandardCharsets.UTF_8);

                        handleEvent(routingKey, message, deliveryTag, channel);
                        //channel.basicAck(deliveryTag, false);
                    }
                });
    }

    private void handleEvent(String eventName, String message, long deliveryTag, Channel channel){
        if(!subscriptionManager.hasSubscriptionForEvent(eventName)) {
            ackMessage(deliveryTag, channel);
        }

        var subs = subscriptionManager.getHandlersForEvent(eventName);

        subs.forEach(subscription -> {
            var result = subscription.handle(message, context);

            if(result.isCancelled()) {
                nackMessage(deliveryTag, channel);
            }else{
                ackMessage(deliveryTag, channel);
            }
        });
    }

    @SneakyThrows
    private void ackMessage(long deliveryTag, Channel channel) {
        channel.basicAck(deliveryTag, false);
    }

    @SneakyThrows
    private void nackMessage(long deliveryTag, Channel channel) {
        channel.basicNack(deliveryTag, false, true);
    }
}
